Add framework docs (#1775)
hashicorp/terraform-website: content/source/docs/plugin/framework/schemas.html.md

---
layout: "extend"
page_title: "Plugin Development - Framework: Schemas"
description: |-
  How to define a schema using the next-generation provider development
  framework.
---

# Schemas

> Terraform providers can specify the constraints of Terraform configuration
> blocks through the use schemas. These schemas define what fields a provider,

Reviewer: ```suggestion
Schemas specify the constraints of Terraform configuration blocks. They define what fields a provider,
```

resource, or data source configuration block has, and give Terraform metadata
about those fields. The schema can be thought of as the "type information" or
the "shape" of a resource, data source, or provider.

> In the next-generation framework, schemas are defined using [the
> `schema.Schema`
> struct](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/schema#Schema)
> and returned from methods on the
> [provider](/docs/plugin/framework/providers.html),
> [resources](/docs/plugin/framework/resources.html) or [data
> sources](/docs/plugin/framework/data-sources.html).

Reviewer: ```suggestion
[The
`schema.Schema`
struct](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/schema#Schema)
defines schemas that are returned from methods on the
[provider](/docs/plugin/framework/providers.html),
[resources](/docs/plugin/framework/resources.html), and [data
sources](/docs/plugin/framework/data-sources.html).
```

## Version

> Every schema has a version. The version is an integer that allows provider
> developers to track changes to their schemas. It is generally only used when

Reviewer: We should address the user as "you" here. We can assume that folks reading this are provider developers. 

```suggestion
Every schema has a version, which is an integer that allows you to track changes to your schemas. It is generally only used when
```

supported](https://github.com/hashicorp/terraform-plugin-framework/issues/42)
in the next-generation framework. At this time, there is no reason to specify a
schema version. It can be safely omitted.

## DeprecationMessage

Not every resource, data source, or provider will be supported forever.
Sometimes designs change or APIs are deprecated. Schemas that have their
`DeprecationMessage` property set will display that message as a warning when
that provider, data source, or resource is used. A good message will tell
practitioners that the provider, resource, or data source is deprecated, and
will indicate a migration strategy.

~> **Important:** Deprecation messages will not currently be surfaced to users,
but [will be in the near
future](https://github.com/hashicorp/terraform-plugin-framework/issues/71).
They can be safely added to your schemas to take advantage of the behavior when
the framework gains this capability.

## Description

Various tooling like
[terraform-plugin-docs](https://github.com/hashicorp/terraform-plugin-docs) and
the [language server](https://github.com/hashicorp/terraform-ls) can use
metadata in the schema to generate documentation or offer a better editor

> experience for practitioners. The `Description` property allows provider
> developers to add a description of a resource, data source, or provider that
> these tools can take advantage of.

Reviewer: ```suggestion
experience for practitioners. Use the `Description` property to add a description of a resource, data source, or provider that these tools can leverage.
```

## MarkdownDescription

Similar to the `Description` property, the `MarkdownDescription` is used to
provide a markdown-formatted version of the description to tooling like
[terraform-plugin-docs](https://github.com/hashicorp/terraform-plugin-docs). It
is a best practice to only alter the formatting, not the content, between the
`Description` and `MarkdownDescription`.

> At the moment, if the `MarkdownDescription` property is set it will always be
> used instead of the `Description` property, but it is possible that in the
> future a different strategy will be employed to surface descriptions to other
> tooling in a format that best suits their needs, and so specifying both is a
> good future-proofing strategy.
> 

Reviewer: ```suggestion
At the moment, if the `MarkdownDescription` property is set it will always be
used instead of the `Description` property. It is possible that a different strategy may be employed in the future to surface descriptions to other tooling in a different format, so we recommend specifying both fields.

```
